# Fase 1: Seleccion y reconocimiento de datos

In [1]:
#libraries
import pandas as pd
import numpy as np

## 1. Analizar el dataset

In [2]:
#import data
df_normal = pd.read_csv('InSDN_DatasetCSV/Normal_data.csv')
df_ovs = pd.read_csv('InSDN_DatasetCSV/OVS.csv')

### 1.1 Tráfico normal

In [3]:
#show features db normal
df_normal.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68424 entries, 0 to 68423
Data columns (total 84 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Flow ID            68424 non-null  object 
 1   Src IP             68424 non-null  object 
 2   Src Port           68424 non-null  int64  
 3   Dst IP             68424 non-null  object 
 4   Dst Port           68424 non-null  int64  
 5   Protocol           68424 non-null  int64  
 6   Timestamp          68424 non-null  object 
 7   Flow Duration      68424 non-null  int64  
 8   Tot Fwd Pkts       68424 non-null  int64  
 9   Tot Bwd Pkts       68424 non-null  int64  
 10  TotLen Fwd Pkts    68424 non-null  float64
 11  TotLen Bwd Pkts    68424 non-null  float64
 12  Fwd Pkt Len Max    68424 non-null  int64  
 13  Fwd Pkt Len Min    68424 non-null  int64  
 14  Fwd Pkt Len Mean   68424 non-null  float64
 15  Fwd Pkt Len Std    68424 non-null  float64
 16  Bwd Pkt Len Max    684

In [4]:
#show statics of db normal
df_normal.describe()

Src Port      Dst Port      Protocol  Flow Duration  Tot Fwd Pkts  \
count  68424.000000  68424.000000  68424.000000   6.842400e+04  68424.000000   
mean   37361.032459   9733.738557     11.418947   1.362636e+07     11.807217   
std    20305.851419  19352.308798      5.551915   3.360107e+07    112.712183   
min        0.000000      0.000000      0.000000  -1.540000e+02      0.000000   
25%    34213.500000     53.000000      6.000000   3.134000e+03      1.000000   
50%    43383.000000    443.000000      6.000000   4.799000e+03      1.000000   
75%    52105.250000    443.000000     17.000000   7.122470e+05      6.000000   
max    65513.000000  60998.000000     17.000000   1.200000e+08  16928.000000   

       Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  \
count  68424.000000     6.842400e+04     6.842400e+04     68424.000000   
mean      18.592628     3.354497e+03     4.128116e+04       425.863644   
std      235.949827     1.560998e+05     7.680108e+05      1425.400448   
min        1.000000     0.000000e+00     0.000000e+00         0.000000   
25%        2.000000     3.000000e+01     1.370000e+02        30.000000   
50%        3.000000     3.600000e+01     5.020000e+02        36.000000   
75%        5.000000     5.990000e+02     7.620000e+02       249.000000   
max    34094.000000     3.160000e+07     1.070000e+08     64239.000000   

       Fwd Pkt Len Min  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count     68424.000000  ...       68424.000000           68424.0   
mean         22.297469  ...           3.891266               0.0   
std          67.228213  ...          57.082340               0.0   
min           0.000000  ...           0.000000               0.0   
25%           0.000000  ...           1.000000               0.0   
50%           0.000000  ...           1.000000               0.0   
75%          34.000000  ...           3.000000               0.0   
max        3900.000000  ...       11180.000000               0.0   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04   
mean   1.887697e+05  1.434025e+05  3.980718e+05  8.563484e+04  4.508770e+06   
std    1.373001e+06  1.105629e+06  2.467808e+06  1.027435e+06  1.314630e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.040000e+08  6.830000e+07  1.040000e+08  1.040000e+08  1.190000e+08   

           Idle Std      Idle Max      Idle Min  
count  6.842400e+04  6.842400e+04  6.842400e+04  
mean   6.041184e+05  5.066966e+06  3.975015e+06  
std    3.486207e+06  1.448106e+07  1.246315e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  
max    7.300000e+07  1.190000e+08  1.190000e+08  

[8 rows x 79 columns]

In [5]:
#encontrar tipos de protocolos
df_normal.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [6]:
#tipo de trafico
df_normal.Label.unique()

array(['Normal'], dtype=object)

**Descripción subconjunto de tráfico normal:**<br>
El subconjunto de tráfico normal cuenta con un total de 84 caracteríticas y 68424 muestras. Los registros de flujo analizados contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object son: 
- Flow ID (contiene la IP de origen y destino, puerto de origen y destino, y protocolo L4).
- Src IP (direccion ip de origen).
- Dst IP (direccion ip de destino). 
- Timestamp: marca de tiempo.
- Label: etiqueta.

La caracteristica Protocolo es de tipo Int y contiene 3 valores; HOPOPT, UDP, TCP, además estos datos ya se encuentran enumerados.
También, existen conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, entre otros. Estos últimos se analizarán con más detalle en la siguiente fase.

### 1.2 Tráfico de ataque

In [7]:
#show features db attacks
df_ovs.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138722 entries, 0 to 138721
Data columns (total 84 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Flow ID            138722 non-null  object 
 1   Src IP             138722 non-null  object 
 2   Src Port           138722 non-null  int64  
 3   Dst IP             138722 non-null  object 
 4   Dst Port           138722 non-null  int64  
 5   Protocol           138722 non-null  int64  
 6   Timestamp          138722 non-null  object 
 7   Flow Duration      138722 non-null  int64  
 8   Tot Fwd Pkts       138722 non-null  int64  
 9   Tot Bwd Pkts       138722 non-null  int64  
 10  TotLen Fwd Pkts    138722 non-null  int64  
 11  TotLen Bwd Pkts    138722 non-null  int64  
 12  Fwd Pkt Len Max    138722 non-null  int64  
 13  Fwd Pkt Len Min    138722 non-null  int64  
 14  Fwd Pkt Len Mean   138722 non-null  float64
 15  Fwd Pkt Len Std    138722 non-null  float64
 16  Bw

In [8]:
#show statics of db attacks
df_ovs.describe()

Src Port       Dst Port       Protocol  Flow Duration  \
count  138722.000000  138722.000000  138722.000000   1.387220e+05   
mean    28590.150805    4058.269092       3.911492   1.705513e+06   
std     23855.062703   11594.755381       2.877929   1.107296e+07   
min         0.000000       0.000000       0.000000   1.000000e+00   
25%         0.000000       0.000000       0.000000   2.000000e+00   
50%     37831.000000      80.000000       6.000000   2.208000e+03   
75%     48394.000000    1058.000000       6.000000   7.661750e+03   
max     63020.000000   65389.000000      17.000000   1.197861e+08   

        Tot Fwd Pkts   Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  \
count  138722.000000  138722.000000    138722.000000    138722.000000   
mean        8.533751       2.950650       149.048759       292.504022   
std      2445.718229      14.918974      1935.634892       741.152741   
min         0.000000       1.000000         0.000000         0.000000   
25%         0.000000       2.000000         0.000000         0.000000   
50%         0.000000       2.000000         0.000000         0.000000   
75%         2.000000       3.000000         0.000000         0.000000   
max    910748.000000    2039.000000     77600.000000     80143.000000   

       Fwd Pkt Len Max  Fwd Pkt Len Min  ...  Fwd Act Data Pkts  \
count    138722.000000    138722.000000  ...      138722.000000   
mean         68.814644         0.026211  ...           1.360224   
std         194.641622         0.962083  ...          47.433114   
min           0.000000         0.000000  ...           0.000000   
25%           0.000000         0.000000  ...           0.000000   
50%           0.000000         0.000000  ...           0.000000   
75%           0.000000         0.000000  ...           0.000000   
max        1448.000000        44.000000  ...        2425.000000   

       Fwd Seg Size Min   Active Mean    Active Std    Active Max  \
count          138722.0  1.387220e+05  1.387220e+05  1.387220e+05   
mean                0.0  5.027155e+04  1.583963e+03  5.242607e+04   
std                 0.0  7.341994e+05  7.711422e+04  7.476735e+05   
min                 0.0  0.000000e+00  0.000000e+00  0.000000e+00   
25%                 0.0  0.000000e+00  0.000000e+00  0.000000e+00   
50%                 0.0  0.000000e+00  0.000000e+00  0.000000e+00   
75%                 0.0  0.000000e+00  0.000000e+00  0.000000e+00   
max                 0.0  6.810000e+07  6.092367e+06  6.810000e+07   

         Active Min     Idle Mean      Idle Std      Idle Max      Idle Min  
count  1.387220e+05  1.387220e+05  1.387220e+05  1.387220e+05  1.387220e+05  
mean   4.941142e+04  1.420137e+06  4.633044e+04  1.468712e+06  1.380748e+06  
std    7.325348e+05  1.024988e+07  7.443023e+05  1.036689e+07  1.020332e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
max    6.810000e+07  8.220000e+07  1.780000e+07  8.220000e+07  8.220000e+07  

[8 rows x 79 columns]

In [9]:
#encontrar tipos de protocolos
df_ovs.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [10]:
# tipos de ataques
df_ovs.Label.unique()

array(['BFA', 'DDoS ', 'DoS', 'Probe', 'Web-Attack', 'BOTNET'],
      dtype=object)

**Descripción subconjunto de tráfico de ataque:**<br>
El subconjunto de tráfico de ataque cuenta con 84 caracteríticas y 138722 muestras. Los tipos de ataque son: fuerza bruta (BFA), DDoS, Dos, Probe, Web y botnet, cuyas muestras están desequilibradas. Así mismo, los registros de flujo contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object siguen siendo las mismas del anterior (normal). Se repiten los mismos protocolos codificados de L4 HOPOPT, UDP y TCP.
También, existen los mismos conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, etc, los cuales serán analizados con más detalle en la siguiente fase.

## 2. Identificación de variables

In [11]:
#concatenar dataset
df_unido = pd.concat([df_normal,df_ovs], axis=0)
#dimension del conjunto total 
df_unido.shape

(207146, 84)

#### 2.1 Matriz de caracteristicas inicial

In [12]:
#seleccion de variable independiente
X = df_unido.iloc[:,:-1].values
print(X)

[['185.127.17.56-192.168.20.133-443-53648-6' '185.127.17.56' 443 ... 0.0
  0.0 0.0]
 ['185.127.17.56-192.168.20.133-443-53650-6' '192.168.20.133' 53650 ...
  0.0 0.0 0.0]
 ['192.168.20.133-192.168.20.2-35108-53-6' '192.168.20.133' 35108 ... 0.0
  0.0 0.0]
 ...
 ['172.17.0.2-200.175.2.130-80-55446-6' '200.175.2.130' 55446 ... 0.0 0.0
  0.0]
 ['172.17.0.2-200.175.2.130-80-55456-6' '200.175.2.130' 55456 ... 0.0 0.0
  0.0]
 ['172.17.0.2-200.175.2.130-80-55456-6' '200.175.2.130' 55456 ... 0.0 0.0
  0.0]]


#### 2.2 Variable objetivo

In [13]:
#seleccion de variable dependiente y nro de cantidad de registros por clase de trafico
y=df_unido.iloc[:,83]
print(f"{y.value_counts()}",f"{y.value_counts().sum()}" )

Normal        68424
DoS           52471
DDoS          48413
Probe         36372
BFA            1110
Web-Attack      192
BOTNET          164
Name: Label, dtype: int64 207146
